### PR TITLE
Space - Hala

### PR DESCRIPTION
**Pull request by halahaddad1**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | A spotlight method, which returns the most voted work to be displayed in the spotlight section.
Describe how you approached testing that model method. What edge cases did you come up with? | I tested that the return of that method is in fact of type work, and that it correctly chooses the work with top votes.
What are `session` and `flash`? What is the difference between them? | Session stores data while other requests are happening and keeps that information to be later used in other requests, where flash is a method that posts temporary messages to the user in response to certain actions.
What was one thing that you gained more clarity on through this assignment? | Session and model testing.
What is the Heroku URL of your deployed application? | https://halas-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/views/works/home.html.erb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            <h2 class="card-title row bold-title">Media Spotlight: <div class="spotlight bold-title"> <%= link_to spotlight.title, work_path(spotlight), style:"text-decoration: underline" %> by <%= spotlight.creator %></div>
+            </h2>
+            <p class="card-text">
+                <p><%= spotlight.votes.count %> votes - <%= spotlight.description %></p>
+
+        </div>
+    </div>
+</div>
+
+<div class="row flex-nowrap">
```
Comment (beccaelenzil) [inline app/views/works/home.html.erb:26]: You essentially have the same code repeated three times. Consider how you could use a partial view or a loop to DRY up this code.

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            <th>Title</th>
+            <th>Created By</th>
+            <th>Published</th>
+            <th>Upvote</th>
+
+        </tr>
+    </thead>
+    <tbody>
+        <h3> Top Albums </h3>
+        <% Work.albums_list.each do |work| %>
```
Comment (beccaelenzil) [inline app/views/works/index.html.erb:26]: As above, you essentially have the same code repeated 3 times. Consider how you may dry up this code, including modifying your model method to take the category as an argument.

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    validates :creator, presence: true
+    validates :publication_year, presence: true
+    validates :description, presence: true
+    CATEGORIES = [:movie,:book,:album]
+
+    def self.spotlight
+        return Work.all.max_by { |work| work.votes.count }
+    end
+
+    def self.movies_list
```
Comment (beccaelenzil) [inline app/models/work.rb:15]: Consider how you could DRY up this code by using the category as an argument.

```
def self.sorted_list(category)
```

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      return
+    else
+      flash[:error] = "Something happened, Media update unsuccessful!"
+      render :edit
+      return
+    end
+  end
+
+  def destroy
+    @work = Work.find_by(id: params[:id])
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:62]: Note that you need to find a work in multiple actions. This would be great functionality to put into a controller filter.

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+      it 'can return a top media item for spotlight' do
+      
+        result = @spotlight_media.valid?
+        
+        expect(result).must_equal true
+        expect(@spotlight_media).must_be_instance_of Work
+        expect(@works.max_by { |work| work.votes.count }).must_equal @spotlight_media
+      end
+
```
Comment (beccaelenzil) [inline test/models/work_test.rb:90]: Consider including another test for what happens if two works are tied.

File: test/models/vote_test.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    expect(@votes.first).wont_be_nil
+  end
+  
+  it "will have all required fields" do
+    expect(@votes.first.user_id).wont_be_nil
+    expect(@votes.first.work_id).wont_be_nil
+  end
+  
+  describe "relationships" do
+    it "has relationships with Work and User" do
```
Comment (beccaelenzil) [inline test/models/vote_test.rb:20]: Consider breaking this into 2 tests since it is 2 different relationships.

**Comment by beccaelenzil** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️?
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️Something's a little funky with the styling/html, and I see the flash message in two places.
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️?
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️I like how you provide this flash message "Media page does not exist! Here's a list of all Media available!"
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️It looks like some extra files were checked into git. Take a look at your gitignore file and make sure everything's there that you expect.

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️?
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | See in-line comments.
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | See in-line comment.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this assignment. Your code is clear and well written, and it is clear that the learning goals were met. Your models are very well tested. I've let a few in-line comments focused on ways you could DRY up your code with partial views and controller filters. Please let me know if you have any questions. Keep up the hard work!
